Mathlib/Data/Finset/Sum.lean

```diff
@@ -94,4 +96,89 @@ theorem disj_sum_strictMono_right (s : Finset α) :
+-/
+def toLeft (s : Finset (α ⊕ β)) : Finset α :=
+  s.disjiUnion (Sum.elim singleton (fun _ => ∅)) <| by
+    simp [Set.PairwiseDisjoint, Set.Pairwise, Function.onFun, eq_comm]
+
+/--
+Given a finset of elements `α ⊕ β`, extract all the elements of the form `β`. This
+forms a quasi-inverse to `disjSum`, in that it recovers its right input.
+-/
+def toRight (s : Finset (α ⊕ β)) : Finset β :=
+  s.disjiUnion (Sum.elim (fun _ => ∅) singleton) <| by
+    simp [Set.PairwiseDisjoint, Set.Pairwise, Function.onFun, eq_comm]
+
+@[simp] lemma mem_toLeft {s : Finset (α ⊕ β)} {x : α} : x ∈ s.toLeft ↔ inl x ∈ s := by
+  simp [toLeft]
+
+@[simp] lemma mem_toRight {s : Finset (α ⊕ β)} {x : β} : x ∈ s.toRight ↔ inr x ∈ s := by
+  simp [toRight]
+
+@[gcongr]
+lemma toLeft_subset_toLeft {s t : Finset (α ⊕ β)} : s ⊆ t → s.toLeft ⊆ t.toLeft :=
+  fun h _ => by simpa only [mem_toLeft] using @h _
+
+@[gcongr]
+lemma toRight_subset_toRight {s t : Finset (α ⊕ β)} : s ⊆ t → s.toRight ⊆ t.toRight :=
+  fun h _ => by simpa only [mem_toRight] using @h _
+
+lemma toLeft_monotone : Monotone (@toLeft α β) := fun _ _ => toLeft_subset_toLeft
+lemma toRight_monotone : Monotone (@toRight α β) := fun _ _ => toRight_subset_toRight
+
+lemma toLeft_disjSum_toRight {s : Finset (α ⊕ β)} : s.toLeft.disjSum s.toRight = s := by
+  ext (x | x) <;> simp
+
+lemma card_toLeft_add_card_toRight {s : Finset (α ⊕ β)} :
+    s.toLeft.card + s.toRight.card = s.card := by
+  rw [← card_disjSum, toLeft_disjSum_toRight]
+
+lemma card_toLeft_le {s : Finset (α ⊕ β)} : s.toLeft.card ≤ s.card :=
+  (Nat.le_add_right _ _).trans_eq card_toLeft_add_card_toRight
+
+lemma card_toRight_le {s : Finset (α ⊕ β)} : s.toRight.card ≤ s.card :=
+  (Nat.le_add_left _ _).trans_eq card_toLeft_add_card_toRight
+
+@[simp] lemma toLeft_disjSum : (s.disjSum t).toLeft = s := by ext x; simp
+
+@[simp] lemma toRight_disjSum : (s.disjSum t).toRight = t := by ext x; simp
+
+lemma disjSum_eq_iff {u : Finset (α ⊕ β)} : s.disjSum t = u ↔ s = u.toLeft ∧ t = u.toRight :=
+  ⟨fun h => by simp [← h], fun h => by simp [h, toLeft_disjSum_toRight]⟩
+
+lemma eq_disjSum_iff {u : Finset (α ⊕ β)} : u = s.disjSum t ↔ u.toLeft = s ∧ u.toRight = t :=
+  ⟨fun h => by simp [h], fun h => by simp [← h, toLeft_disjSum_toRight]⟩
+
+variable [DecidableEq α] [DecidableEq β] {s t : Finset (α ⊕ β)}
+
+lemma toLeft_image_swap {s : Finset (α ⊕ β)} : (s.image Sum.swap).toLeft = s.toRight := by
+  ext x; simp
+
+lemma toRight_image_swap {s : Finset (α ⊕ β)} : (s.image Sum.swap).toRight = s.toLeft := by
+  ext x; simp
+
+@[simp] lemma toLeft_insert_inl : (insert (inl a) s).toLeft = insert a s.toLeft := by ext y; simp
+@[simp] lemma toLeft_insert_inr : (insert (inr b) s).toLeft = s.toLeft := by ext y; simp
+@[simp] lemma toRight_insert_inl : (insert (inl a) s).toRight = s.toRight := by ext y; simp
+@[simp] lemma toRight_insert_inr : (insert (inr b) s).toRight = insert b s.toRight := by ext y; simp
```

Comment:
Can you add the `cons` version of these too?

Reply:
Done